From be69f897102f653ffacd9ebf73af44ce6cf5319c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sat, 5 Mar 2022 20:52:30 +0100
Subject: [PATCH 05/19] winex11.drv: Factor focus window lookups together in
 find_activatable_window.

CW-Bug-Id: #19553
CW-Bug-Id: #19644
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=2155
---
 dlls/winex11.drv/event.c | 75 +++++++++++++++++++---------------------
 1 file changed, 36 insertions(+), 39 deletions(-)

diff --git a/dlls/winex11.drv/event.c b/dlls/winex11.drv/event.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/event.c
+++ b/dlls/winex11.drv/event.c
@@ -567,10 +567,41 @@ static void set_input_focus( struct x11drv_win_data *data )
 
 }
 
+
+static HWND find_activatable_window( HWND hwnd, BOOL check )
+{
+    LRESULT result;
+
+    if (can_activate_window( hwnd ))
+    {
+        if (!check) return hwnd;
+
+        /* simulate a mouse click on the menu to find out
+         * whether the window wants to be activated */
+        result = send_message( hwnd, WM_MOUSEACTIVATE, (WPARAM)NtUserGetAncestor( hwnd, GA_ROOT ),
+                               MAKELONG( HTMENU, WM_LBUTTONDOWN ) );
+        if (result != MA_NOACTIVATEANDEAT && result != MA_NOACTIVATE) return hwnd;
+    }
+
+    if (hwnd == NtUserGetDesktopWindow()) hwnd = NtUserGetForegroundWindow();
+    else
+    {
+        hwnd = get_focus();
+        if (hwnd) hwnd = NtUserGetAncestor( hwnd, GA_ROOT );
+        if (!hwnd) hwnd = get_active_window();
+    }
+
+    if (!hwnd) hwnd = x11drv_thread_data()->last_focus;
+    if (!hwnd) hwnd = NtUserGetDesktopWindow();
+
+    if (!can_activate_window( hwnd )) return 0;
+    return hwnd;
+}
+
 /**********************************************************************
  *              set_focus
  */
-static void set_focus( Display *display, HWND hwnd, Time time )
+static void set_focus( Display *display, HWND hwnd, Time time, BOOL check )
 {
     HWND focus, old_active;
     Window window;
@@ -578,7 +609,8 @@ static void set_focus( Display *display, HWND hwnd, Time time )
 
     old_active = NtUserGetForegroundWindow();
 
-    TRACE( "setting foreground window to %p\n", hwnd );
+    if (!(focus = find_activatable_window( hwnd, check ))) return;
+    TRACE( "setting foreground window to %p\n", focus );
     NtUserSetForegroundWindow( hwnd );
 
     /* Some applications expect that a being deactivated topmost window
@@ -697,33 +729,7 @@ static void handle_wm_protocols( HWND hwnd, XClientMessageEvent *event )
                (int)NtUserGetWindowLongW(hwnd, GWL_STYLE),
                get_focus(), get_active_window(), NtUserGetForegroundWindow(), last_focus );
 
-        if (can_activate_window(hwnd))
-        {
-            /* simulate a mouse click on the menu to find out
-             * whether the window wants to be activated */
-            LRESULT ma = send_message( hwnd, WM_MOUSEACTIVATE,
-                                       (WPARAM)NtUserGetAncestor( hwnd, GA_ROOT ),
-                                       MAKELONG( HTMENU, WM_LBUTTONDOWN ) );
-            if (ma != MA_NOACTIVATEANDEAT && ma != MA_NOACTIVATE)
-            {
-                set_focus( event->display, hwnd, event_time );
-                return;
-            }
-        }
-        else if (hwnd == NtUserGetDesktopWindow())
-        {
-            hwnd = NtUserGetForegroundWindow();
-            if (!hwnd) hwnd = last_focus;
-            if (!hwnd) hwnd = NtUserGetDesktopWindow();
-            set_focus( event->display, hwnd, event_time );
-            return;
-        }
-        /* try to find some other window to give the focus to */
-        hwnd = get_focus();
-        if (hwnd) hwnd = NtUserGetAncestor( hwnd, GA_ROOT );
-        if (!hwnd) hwnd = get_active_window();
-        if (!hwnd) hwnd = last_focus;
-        if (hwnd && can_activate_window(hwnd)) set_focus( event->display, hwnd, event_time );
+        set_focus( event->display, hwnd, event_time, TRUE );
     }
     else if (protocol == x11drv_atom(_NET_WM_PING))
     {
@@ -799,16 +805,7 @@ static BOOL X11DRV_FocusIn( HWND hwnd, XEvent *xev )
 
     if (use_take_focus) return TRUE;
 
-    if (!can_activate_window(hwnd))
-    {
-        HWND hwnd = get_focus();
-        if (hwnd) hwnd = NtUserGetAncestor( hwnd, GA_ROOT );
-        if (!hwnd) hwnd = get_active_window();
-        if (!hwnd) hwnd = x11drv_thread_data()->last_focus;
-        if (hwnd && can_activate_window(hwnd)) set_focus( event->display, hwnd, CurrentTime );
-        return TRUE;
-    }
-    else set_focus( event->display, hwnd, CurrentTime );
+    set_focus( event->display, hwnd, CurrentTime, FALSE );
     return TRUE;
 }
 
-- 
0.0.0

